rts/Lua/LuaMatrix.hpp

```diff
@@ -0,0 +1,197 @@
+
+	void Ortho(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+	void Frustum(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+	void Billboard(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+
+	tuple16f GetAsScalar()
+	{
+		return std::make_tuple(
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		);
+	}
+
+	sol::as_table_t<std::vector<float>> GetAsTable()
+	{
+		return sol::as_table(std::vector<float> {
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		});
+	}
+
+public:
+	const CMatrix44f& GetMatRef() const { return  mat; }
+	const CMatrix44f* GetMatPtr() const { return &mat; }
+public:
+	sol::as_table_t<std::vector<float>> operator* (const sol::table& tbl) const {
+		float4 f4 { 0.0f, 0.0f, 0.0f, 1.0f };
+		for (auto i = 1; i <= 4; ++i) {
+			const sol::optional<float> maybeValue = tbl[i];
+			if (!maybeValue)
+				continue;
+
+			f4[i - 1] = maybeValue.value();
+		}
+		f4 = mat * f4;
+		return sol::as_table( std::vector<float> { f4.x, f4.y, f4.z, f4.w } );
+	};
+	LuaMatrixImpl operator* (const LuaMatrixImpl& lmi) const { return LuaMatrixImpl(mat * lmi.GetMatRef()); };
+
+	LuaMatrixImpl operator+ (const LuaMatrixImpl& lmi) const { return LuaMatrixImpl(mat + lmi.GetMatRef()); };
+private:
+	template<typename TFunc>
+	inline void AssignOrMultMatImpl(const sol::optional<bool> mult, const bool multDefault, const TFunc& tfunc);
+
+	template<typename TObj>
+	inline static const TObj* ParseObject(int objID, sol::this_state L);
+
+	template<typename TObj>
+	inline bool GetObjectMatImpl(const unsigned int objID, CMatrix44f& outMat, sol::this_state L);
+	template <typename  TObj>
+	bool ObjectMatImpl(const unsigned int objID, bool mult, sol::this_state L);
+
+	template <typename  TObj>
+	bool GetObjectPieceMatImpl(const unsigned int objID, const unsigned int pieceNum, CMatrix44f& outMat, sol::this_state L);
+	template <typename  TObj>
+	bool ObjectPieceMatImpl(const unsigned int objID, const unsigned int pieceNum, bool mult, sol::this_state L);
+
+	template <typename  TObj>
+	inline static bool IsObjectVisible(const TObj* obj, sol::this_state L);
+
+private:
+	static void SetupScreenMatrices();
+
+	inline static const LocalModelPiece* ParseObjectConstLocalModelPiece(const CSolidObject* obj, const unsigned int pieceNum);
+	//static methods
+private:
+	static constexpr bool viewProjMultDefault = false;
+	static constexpr bool objectMultDefault = true;
+	//static constants
+private:
+	CMatrix44f mat;
+private:
+	static CMatrix44f screenViewMatrix;
+	static CMatrix44f screenProjMatrix;
+	//private vars
+};
+
+class LuaMatrix {
```

Comment:
Since you're already splitting LuaMatrix into two classes (interface and implementation), it might be a good idea to also fully move LuaMatrixImpl to a different file and only include it in the LuaMatrix.cpp and LuaMatrixImpl.cpp files.
This will reduce the amount of stuff that gets included when we include LuaMatrix.h, and should make compilation faster.